linode/manager
# refactor: [M3-8101] - Use `network-transfer/prices` endpoint for transfer overage pricing

## Description 📝
This PR is a follow up to #10468 and uses the newly added `network-transfer/prices` endpoint to fetch network transfer overage pricing from the API, finishing off the transition for OBJ to dynamic pricing data. 

Loading and error states are addressed, especially in the OMC Create Bucket drawer, which I missed in the first PR - see Reproduction steps in the Testing section.

## Changes  🔄
- Created the necessary api-v4 endpoint, query, and interface to use the  `network-transfer/prices` endpoint.
- Created the necessary factories and endpoint in the MSW to support a mocked response to the API endpoint.
 - Added UX for error handling in the unlikely event that network transfer pricing cannot be calculated: disable primary action button and display tooltip error.
- Added loading state for network transfer pricing.
- Updated unit tests.

## Target release date 🗓️
6/24

## Preview 📷

| Before  | After   |
| ------- | ------- |
| ![Screenshot 2024-06-11 at 9 24 00 AM](https://github.com/linode/manager/assets/114685994/2eb7609d-30c8-461c-ac86-26e9dea3b64a) ![Screenshot 2024-06-11 at 9 24 11 AM](https://github.com/linode/manager/assets/114685994/2e05ad73-4027-4e93-a30c-8766d6df1e9c) ![Screenshot 2024-06-11 at 9 24 25 AM](https://github.com/linode/manager/assets/114685994/1d12441f-5349-4124-8b18-c8ad2fdc512d) | ![Screenshot 2024-06-11 at 9 20 17 AM](https://github.com/linode/manager/assets/114685994/6897be45-f728-4974-ad4b-95b88e82cb34) ![Screenshot 2024-06-11 at 9 20 27 AM](https://github.com/linode/manager/assets/114685994/f459d57e-7572-4c9e-a27a-7a857444001a) ![Screenshot 2024-06-11 at 9 20 37 AM](https://github.com/linode/manager/assets/114685994/003bbc88-54b1-4d26-81bb-f549b18c3c32)| 
| n/a | ![Screenshot 2024-06-11 at 9 21 03 AM](https://github.com/linode/manager/assets/114685994/8bb23255-0391-48a6-b927-3837684944ca) |
| n/a | ![Screenshot 2024-06-11 at 9 22 12 AM](https://github.com/linode/manager/assets/114685994/ae367aba-11a0-4a4d-87df-6c3ebccad9c1) |

## How to test 🧪

### Prerequisites


- For OMC, ensure you have the account capability with the necessary tags on your prod account. We are depending on this and not the LD flag. 

### Reproduction steps
(How to reproduce the issue, if applicable)
- To observe that we were missing loading and error state in the OMC Create Bucket drawer, use the React Query dev tools to "Trigger loading" and "Trigger error" states for the [ "network-transfer", "prices"] and ["object-storage", "types"] endpoints. Observe the primary action button does not have loading or error state.

### Verification steps
(How to verify changes)
- Test the following with the OMC feature enabled (tags on account) and disabled (no OMC tags on account):
   - Go to http://localhost:3000/object-storage/buckets/create.
   - Select a regularly priced region (e.g. us-east), Jakarta, and Sao Paulo. For each, confirm that outbound transfer overage pricing (the second price) matches production.
   - Confirm you can still create a bucket. 
   - Confirm that loading and error states are present by toggling in RQ dev tools as described above, or block the network requests. 

## As an Author I have considered 🤔

*Check all that apply*
